### ISSUE TITLE
MacOS github runner deprecated

### ISSUE THREAD

**Issue by NikoOinonen** [Probe-Particle/ppafm#288]
The workflow for publishing the PyPI package currently uses the `macos-11` runner for making the MacOS build. It looks like this was just recently deprecated: https://github.blog/changelog/2024-05-20-actions-upcoming-changes-to-github-hosted-macos-runners/. The publishing workflow is now just hanging forever.

There are several replacing runners: `macos-12`, `macos-13`, `macos-14`, and `macos-latest`. I don't use Mac so I don't know what the differences between the versions are. @yakutovicha I think you were using ppafm on a Mac. Do you have any opinion on this?

**Comment by NikoOinonen** [comment]
I tested and it seems to build on `macos-latest`. I cannot test whether the built wheel actually works, so I'll just assume it's fine. I'll make a PR with this.

**Comment by yakutovicha** [comment]
Unfortunately, the build didn't work. The downloaded package didn't contain the '*.so' files.

The latest version of the application that worked for me (Mac, Apple M1 Max chip) is 0.2.2. The installation, however, does the compilation of the code directly on my computer.

**Comment by ondrejkrejci** [comment]
Confirmed on maxOS Tahore v. 26.4.1 .
Things are even more complicated as `-fopenmp` flag is creating problems, for direct compilation - my way around (in conda) is to `conda install -c conda-forge clang llvm-openmp` and then used the latest version of `clang++` instead of `g++`. (I have create a soft-link in my env bin directoty). As this is pretty cumbersome, so I was thinking, if we could create a conda package maybe? @yakutovicha or @NikoOinonen - do you have some experience in this?

**Comment by NikoOinonen** [comment]
I don't know about conda packages, but if it works with clang, then maybe we can just change the compiler to clang for the automated macOS build.

**Comment by ondrejkrejci** [comment]
Nope, it cannot be done easily. The problem is that `clang++` is automatically installed on mac, and `conda install -c conda-forge clang llvm-openmp` is installing just a latest version of it. For me it was `clang++-22`. Using `clang` only cannot create the shared libraries.

**Comment by NikoOinonen** [comment]
We can also use conda environments on the Github runners for compiling the binaries.

**Comment by NikoOinonen** [comment]
> `-fopenmp` flag is creating problems

I'm also a bit curious as to what the problem actually is. Maybe our code is just written badly that it does not compile. `gcc` tends to allow for more non-standard stuff than `clang` from what I recall.

**Comment by ondrejkrejci** [comment]
And openCL is a complete mess ... 
In bash:
```
export PPAFM_OCL_DIR=/opt/miniconda3/envs/ppafm/lib/python3.12/site-packages/ppafm/ocl/cl
```
Inside python:
```
- #include "splines.cl"
+ #include <splines.cl>
```
Rerun → **should** compile on CPU only.

I did not go that far, to change the code itself to try. It is maybe a question, whether we continue with OCL development, or we should maybe try ways, how to compile c++ directly on GPUs.

**Comment by ondrejkrejci** [comment]
> > `-fopenmp` flag is creating problems
> 
> I'm also a bit curious as to what the problem actually is. Maybe our code is just written badly that it does not compile. `gcc` tends to allow for more non-standard stuff than `clang` from what I recall.

Yes, it seems that the simplest way to go around it would be to not to use `-fopenmp` flag on mac. I have compiled my `g++` on my mac, when I put there _conda_ through _homebrew_ but that is just an old `clang` which cannot use openmp.

**Comment by ondrejkrejci** [comment]
And by the way `conda install conda-forge::gcc` and `conda install conda-forge::gxx` cannot be used as well, as then g++ libraries are not included.

**Comment by NikoOinonen** [comment]
> an old `clang` which cannot use openmp

From a quick search, it looks like `clang`  has supported OpenMP for at least 10 years, so I would be surprised if the compiler is that old. Might be that it does not support the specific version of the OpenMP standard that we rely on?

For the OCL compilation, we do it on-the-fly as the program runs using `pyopencl`. The relevant part is here: https://github.com/Probe-Particle/ppafm/blob/40f309be26ace1cc91553ad3c47a4afc370d27bd/ppafm/ocl/oclUtils.py#L23-L31
If it needs some additional flags on mac, we can add it there.

**Comment by ondrejkrejci** [comment]
> > an old `clang` which cannot use openmp
> 
> From a quick search, it looks like `clang` has supported OpenMP for at least 10 years, so I would be surprised if the compiler is that old. Might be that it does not support the specific version of the OpenMP standard that we rely on?

```
% g++ --version
Apple clang version 21.0.0 (clang-2100.0.123.102)
Target: arm64-apple-darwin25.4.0
```

So it is not about the version itself, but it is also connected with the libraries linkage. I have tried this: ` cpp % g++ -fPIC -std=c++11 -O2 -mtune=native -Xpreprocessor -fopenmp -lomp -c ProbeParticle.cpp -o ProbeParticle.o`
But it failed with:
```
ProbeParticle.cpp:11:10: fatal error: 'omp.h' file not found
   11 | #include <omp.h>
      |          ^~~~~~~
4 warnings and 1 error generated.
```
So, g++ itself should work, but without openmp.

**Comment by ondrejkrejci** [comment]
> For the OCL compilation, we do it on-the-fly as the program runs using `pyopencl`. The relevant part is here:
> 
> [ppafm/ppafm/ocl/oclUtils.py](https://github.com/Probe-Particle/ppafm/blob/40f309be26ace1cc91553ad3c47a4afc370d27bd/ppafm/ocl/oclUtils.py#L23-L31)
> 
> Lines 23 to 31 in [40f309b](/Probe-Particle/ppafm/commit/40f309be26ace1cc91553ad3c47a4afc370d27bd)
> 
>  def loadProgram(self, fname): 
>      cl_path = str(self.CL_PATH) 
>      if self.platform.name != "Portable Computing Language": 
>          # Older versions of pocl don't handle quotes and spaces properly. This is kind of ugly, but 
>          # this is needed for the version of pocl running on Github Actions at the moment of writing. 
>          cl_path = f'"{cl_path}"' 
>      with open(fname) as f: 
>          program = cl.Program(self.ctx, f.read()).build(options=["-I", cl_path]) 
>      return program

I have tried to run the OCL graphene `python run_gpu.py` example, but anyway, I was not able to link the path properly. Maybe it would work if we would do the change of include:
```
- #include "splines.cl"
+ #include <splines.cl>
```
But I do not know, if this would change the behaviour in linux or windows.

**Comment by NikoOinonen** [comment]
At least on Linux changing to `<splines.cl>` works without any other changes.

**Comment by ondrejkrejci** [comment]
Ok, I will try to unset the PATH and change it later. Thanks @NikoOinonen!

**Comment by ondrejkrejci** [comment]
So I have been looking at this a little bit, since I am also compiling PPSTM on my mac laptop. This seems to be working:
`conda env create -f env_mac.yml `
with
`env_mac.yml`being:
````
name: ppafm_mac

channels:
  - conda-forge

dependencies:
  - clangxx_osx-arm64=22
  - llvm-openmp=22
  - make=4.4.1
  - python=3.12.13
````
and then the CC compiler should be: `clang++-22` and then `make` works just fine.

**Comment by ondrejkrejci** [comment]
Hi, both - connected directly to the runners: Could this help?
https://github.blog/changelog/2026-02-26-macos-26-is-now-generally-available-for-github-hosted-runners/
